tls: add support for string matcher for subject alt name validation

ramaraochavali opened the pull request:
Description: Adds support for generic StringMatcher for subject alt name validation.
Risk Level: Low
Testing: Added
Docs Changes: Updated
Release Notes: Updated
Fixes https://github.com/envoyproxy/envoy/issues/7813

lizan commented on include/envoy/ssl/certificate_validation_context_config.h:
```diff
@@ -40,6 +41,12 @@ class CertificateValidationContextConfig {
+  /**
+   * @return The subject alt name matchers to be verified, if enabled. Otherwise, ""
+   */
+  virtual const std::vector<::envoy::type::matcher::StringMatcher>
```
return const vector&?

ramaraochavali replied:
yeah. sorry missed it. will change

lizan commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -567,6 +586,66 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+  if (san_names == nullptr) {
+    return false;
+  }
+  for (const GENERAL_NAME* san : san_names.get()) {
```
can you extract an utility function to convert GENERAL_NAME to string? then this for will be simpler.

lizan commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -567,61 +620,35 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
-          return true;
-        }
+    for (auto& config_san_matcher : match_subject_alt_name_list) {
+      if (config_san_matcher.match(generalNameAsString(san))) {
```
store result of `generalNameAsString(san)` outside of the inner loop.

ramaraochavali replied:
Yeah. I noticed that and fixed it. Next commit will have it

jmarantz commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -567,61 +620,35 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+  for (const GENERAL_NAME* san : san_names.get()) {
+    for (auto& config_san : subject_alt_names) {
+      if (san->type == GEN_DNS ? dnsNameMatch(config_san, generalNameAsString(san).c_str())
+                               : config_san == generalNameAsString(san))
```
make a temp for generalNameAsString(san), hoisting to the body of the outer-loop.

ramaraochavali commented on api/envoy/api/v3alpha/auth/cert.proto:
```diff
@@ -279,15 +284,15 @@ message CertificateValidationContext {
   //   therefore this option must be used together with :ref:`trusted_ca
   //   <envoy_api_field_api.v3alpha.auth.CertificateValidationContext.trusted_ca>`.
-  repeated string verify_subject_alt_name = 4;
+  repeated type.matcher.v3alpha.StringMatcher match_subject_alt_name = 9;
```
Once we deprecate in v2, the tooling will automatically remove it from v3.

lizan replied:
@jmarantz this is in v3alpha so we're good with it.

ramaraochavali commented:
@lizan addressed all comments, tests are passing. There are few example configs + doc + very few tests that still use `verify_subject_alt_name` - I think they can be converted to new config when the field is entirely removed. PTAL.

ramaraochavali commented:
@lizan @PiotrSikora can you PTAL when you get some time?

lizan commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -567,61 +619,36 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+  for (const GENERAL_NAME* general_name : san_names.get()) {
+    const std::string san = generalNameAsString(general_name);
+    for (auto& config_san : subject_alt_names) {
+      if (general_name->type == GEN_DNS ? dnsNameMatch(config_san, san.c_str()) : config_san == san)
```
always `{}`. erm this should be caught by clang-tidy, cc @derekargueta

lizan replied:
seems clang-tidy fails but azp is not failing the job. I'm investigating

ramaraochavali replied:
great catch. Don't know how I missed it.

ramaraochavali commented:
@lizan fixed the clang-tidy issue, PTAL

ramaraochavali commented:
@lizan when you get chance, can you PTAL?

ramaraochavali commented:
@lizan can you PTAL when you get chance?

@htuch @mattklein123 can you please check from api perspective if this looks good?

mattklein123 commented on include/envoy/ssl/certificate_validation_context_config.h:
```diff
@@ -40,6 +41,12 @@ class CertificateValidationContextConfig {
   virtual const std::vector<std::string>& verifySubjectAltNameList() const PURE;
 
+  /**
+   * @return The subject alt name matchers to be verified, if enabled. Otherwise, ""
```
nit: `Otherwise, ""` doesn't make sense.

mattklein123 commented on include/envoy/ssl/certificate_validation_context_config.h:
```diff
@@ -40,6 +41,12 @@ class CertificateValidationContextConfig {
+   * @return The subject alt name matchers to be verified, if enabled. Otherwise, ""
+   */
+  virtual const std::vector<::envoy::type::matcher::StringMatcher>&
+  matchSubjectAltNameList() const PURE;
```
nit: since you are editing this code, I would probably call this `subjectAltNameMatchers()` or something like that?

mattklein123 commented on test/server/listener_manager_impl_test.cc:
```diff
@@ -130,7 +130,7 @@ per_connection_buffer_limit_bytes: 8192
 }
 
-TEST_F(ListenerManagerImplWithRealFiltersTest, TlsTransportSocket) {
+TEST_F(ListenerManagerImplWithRealFiltersTest, DEPRECATED_FEATURE_TEST(TlsTransportSocket)) {
```
test nit: I would fix this test, since it's really not that deprecated behavior you care about. Instead, I would have a specific TLS test that tests the old deprecated behavior and thus a DEPRECATED_FEATURE_TEST there? WDYT?

ramaraochavali replied:
Yeah. That makes. I will fix it. We already have a TLSContext test that I marked deprecated.

ramaraochavali commented:
@mattklein123 Thanks for the review. I addressed your comments. If you are good with API, can you please approve the api part?

@lizan PTAL, when you get chance.

htuch commented on api/envoy/api/v2/auth/cert.proto:
```diff
@@ -265,7 +266,17 @@ message CertificateValidationContext {
+  //   Subject Alternative Names are easily spoofable and verifying only them is insecure,
+  //   therefore this option must be used together with :ref:`trusted_ca
+  //   <envoy_api_field_auth.CertificateValidationContext.trusted_ca>`.
+  repeated type.matcher.StringMatcher match_subject_alt_name = 9;
```
Nit: this should be a pluralized name as per our proto style guide.

ramaraochavali replied:
Ok. Changed. PTAL.

htuch commented:
/lgtm api

ramaraochavali commented:
/retest

ramaraochavali commented:
@lizan api has been approved. PTAL when you get chance.

ramaraochavali commented:
@lizan  Thank you.

@mattklein123 can you PTAL when you get chance if it needs any other changes?

mattklein123 commented on test/extensions/transport_sockets/tls/context_impl_test.cc:
```diff
@@ -62,6 +62,16 @@ TEST_F(SslContextImplTest, TestVerifySubjectAltNameDNSMatched) {
   EXPECT_TRUE(ContextImpl::verifySubjectAltName(cert.get(), verify_subject_alt_name_list));
 }
 
+TEST_F(SslContextImplTest, TestMatchSubjectAltNameDNSMatched) {
```
Where are you testing the old behavior now? I don't see any new deprecated feature tests?

ramaraochavali replied:
New tests I have added in this class to test the new behaviour. For config test, this test https://github.com/envoyproxy/envoy/blob/master/test/server/listener_manager_impl_test.cc#L174 which is already deprecated, tests the old behaviour. Am I missing some thing else?

ramaraochavali replied:
I have added a new test for `TlsTransportSocket` as well that runs with old config, if that is what you were looking for. PTAL

ramaraochavali commented on test/extensions/transport_sockets/tls/context_impl_test.cc:
```diff
@@ -62,6 +62,16 @@ TEST_F(SslContextImplTest, TestVerifySubjectAltNameDNSMatched) {
+  EXPECT_TRUE(ContextImpl::matchSubjectAltName(cert.get(), subject_alt_name_matchers));
+}
+
 TEST_F(SslContextImplTest, TestVerifySubjectAltNameURIMatched) {
```
@mattklein123 I have added new tests to check the new behaviour. Old tests still exists. Here is one for URI match. Here deprecation feature test is not needed because it is not using the config directly.

ramaraochavali commented on test/extensions/transport_sockets/tls/context_impl_test.cc:
```diff
@@ -62,6 +62,16 @@ TEST_F(SslContextImplTest, TestVerifySubjectAltNameDNSMatched) {
   EXPECT_TRUE(ContextImpl::verifySubjectAltName(cert.get(), verify_subject_alt_name_list));
```
Here is another one.

ramaraochavali commented on test/extensions/transport_sockets/tls/context_impl_test.cc:
```diff
@@ -70,13 +80,33 @@ TEST_F(SslContextImplTest, TestVerifySubjectAltNameURIMatched) {
+  EXPECT_TRUE(ContextImpl::matchSubjectAltName(cert.get(), subject_alt_name_matchers));
+}
+
 TEST_F(SslContextImplTest, TestVerifySubjectAltNameNotMatched) {
```
here is another old one
